README.md
# NuxtPress - WordPress x Nuxt Development Environment

<p align="center">
  <img width="300" height="300" src="https://user-images.githubusercontent.com/2754728/74832907-462a9780-5310-11ea-9cc7-9a581fcde91d.png">
</p>

**NuxtPress** combines [**_WordPress_**](https://wordpress.com/), the worlds biggest CMS with [**_NuxtJS_**](https://nuxtjs.org/),
a brilliant front-end application framework, built on top of [**_Vue_**](https://vuejs.org/v2/guide/).

This README contains the quick start and intro documentation.  
Check out [the documentation directory](docs) for more detailed information on NuxtPress.

- [Quick Start](#quick)
- [Introduction](#intro)
- [Getting started](docs/1-GettingStarted.md#start)
  - [With Docker](docs/1-GettingStarted.md#with-docker)
    - [Setup WordPress](docs/1-GettingStarted.md#setup-wp)
    - [Setup Nuxt.js](docs/1-GettingStarted.md#setup-nuxt-docker)
    - [Custom container configuration](docs/1-GettingStarted.md#custom-container-configuration)
  - [Without Docker](docs/1-GettingStarted.md#without-docker)
    - [Setup WordPress](docs/1-GettingStarted.md#setup-wp-no-docker)
    - [Setup Nuxt.js](docs/1-GettingStarted.md#setup-nuxt-no-docker)
    - [URL and Port Configuration](docs/1-GettingStarted.md#port-config)
- [WordPress Rest API endpoints](docs/2-WordPressRestApi.md#ep)
  - [Extensions to the API endpoints](docs/2-WordPressRestApi.md#epp)
    - [Front-page](docs/2-WordPressRestApi.md#epp-front)
    - [Menus](docs/2-WordPressRestApi.md#epp-menu)
    - [Slugs](docs/2-WordPressRestApi.md#epp-slugs)
    - [Meta queries](docs/2-WordPressRestApi.md#epp-meta)
    - [Taxonomy queries](docs/2-WordPressRestApi.md#epp-tax)
    - [Geo queries](docs/2-WordPressRestApi.md#epp-geo)
    - [Custom post types](docs/2-WordPressRestApi.md#epp-cpt)
- [Development Tools](docs/3-DevelopmentTools.md#development-tools)
  - [StoryBook](docs/3-DevelopmentTools.md#storybook)
  - [Jest (Unit + Snapshot Testing)](docs/3-DevelopmentTools.md#jest)
  - [SCSS](docs/3-DevelopmentTools.md#scss)
  - [Lint Staged Files on Commit](docs/3-DevelopmentTools.md#lint-staged)
  - [Scripts](docs/3-DevelopmentTools.md#scripts)
    - [Working with the containers](docs/3-DevelopmentTools.md#scripts-containers)
      - [WP-CLI and yarn](docs/3-DevelopmentTools.md#scripts-containers-tools)
    - [Scaffolding](docs/3-DevelopmentTools.md#scripts-scaffolding)
- [Generate and Deploy](docs/4-GenerateStaticSite.md#deploy)
- [Links](#links)
- [Credits](#cred)

## Quick start

<a name="quick"/>

    git clone https://github.com/hex-digital/nuxtpress.git
    cd nuxtpress
    docker-compose up -d wp.nuxtpress
    cd nuxt && cp .env.dist .env && cp assets/scss/settings/_settings.config.dist.scss assets/scss/settings/_settings.config.scss
    yarn install && yarn dev

- [http://localhost:3080/install.php](http://localhost:3080/install.php) - Install WordPress
- [http://localhost:3080/wp-admin/options-permalink.php](http://localhost:3080/wp-admin/options-permalink.php) - Set permalinks to _Post name_
- [http://localhost:3080/wp-admin/themes.php](http://localhost:3080/wp-admin/themes.php) - Activate **nuxtpress**-theme
- [http://localhost:3000](http://localhost:3000) - Done

## Introduction

<a name="intro"/>

NuxtPress aims to provide a ready to use development environment, combining WordPress with Nuxt.
Included are:

- Fully dockerized **_WordPress_** and **_NuxtJS_** containers, setup in one command: `docker-compose up -d`

- Extended Rest API gives easy access to meta-fields, featured media menus and front-page config.

- Latest **_NuxtJS_**, with a WordPress `$wp` object, to connect to the extended **_WordPress_** Rest API.

This allows you to skip all the backend coding, but provides the flexibility to extend the backend if you want to.

## Showcase

<a name="showcase"/>

The following sites are powered by NuxtPress:

[https://prostack.uk](https://prostack.uk)

Get in touch to have your site listed here.

## Links

<a name="links"/>

[WUXT Headless WordPress API Extensions](https://wordpress.org/plugins/wuxt-headless-wp-api-extensions/): Plugin which includes all our API extensions.

## Credits

<a name="cred"/>

[@northosts](https://github.com/northosts) for the fantastic Wuxt framework on which NuxtPress is hugely dependent. [https://github.com/northosts/wuxt](https://github.com/northosts/wuxt)  
[@yashha](https://github.com/yashha/wp-nuxt/commits?author=yashha) for the excellent idea with the `$wp` object, first implemented in [https://github.com/yashha/wp-nuxt](https://github.com/yashha/wp-nuxt)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### 🚨 Breaking Changes
- Move wp-content into a top level 'wordpress' directory [#13](https://github.com/hex-digital/nuxtpress/issues/13)

### 🐛 Bug Fixes 
- Fix some incorrect naming related to the Typography components [#11](https://github.com/hex-digital/nuxtpress/pull/11)
- Fix a bug where the FlexibleContent component failed when this.post or this.post.acf was undefined [#11](https://github.com/hex-digital/nuxtpress/pull/11)

### 🚀 Features  
- Add support for WordPress templates [#6](https://github.com/hex-digital/nuxtpress/issues/6)
- Add example login page and integrate auth store with WP API client [#7](https://github.com/hex-digital/nuxtpress/issues/7)

### 🏡 Chore
- Add missing environment files to the README and docs  

## [0.0.2] - 2020-02-15

### 🌟 Highlights
- Now using Nuxt 2.11
- Docker compose is now v3, and the mysql container now uses MariaDB v10.4.12.

### 🐛 Bug Fixes 
- Fix undefined key on homepage when getting an error in post retrieval.

### 🦄️ Improvements
- Improved the docker-compose file for easier use, such as naming the volumes.
- Improve the lint-staged files by adding relevant tests and stylelint automatically
- Made the TemporaryPostData helper component only display in development mode

### 🏡 Chore
- Add some missing files to .gitignore, including WP's debug.log   

### ⚓ Dependency Upgrades 
- Upgrade Nuxt from 2.10 to 2.11

## [0.0.1] - 2019-12-03
### 🌟 Highlights

- Initial release of NuxtPress
  - Docker compose for quickly starting development
  - Page components that help you automatically map WordPress API calls to your pages
  - Recommend plugins included from installation
  - Recommended development tools pre-setup, lik Jest, ESLint, Prettier, Stylelint, lint-staged, storybook and more 

For reference:

🌟 Highlights
🚨 Breaking Changes  
🐛 Bug Fixes  
🚀 Features  
🦄️ Improvements
💅 Refactors  
🔥 Performance  
📝 Examples  
🏡 Chore  
👓 Tests  
⚓ Dependency Upgrades  
💖 Thanks to  


nuxt/nuxt.config.js
import axios from 'axios';
import { build } from './config';
require('dotenv').config();

const debug = process.env.NUXTPRESS_NUXT_DEBUG ? process.env.NUXTPRESS_NUXT_DEBUG : false;
const wpUrl = process.env.NUXTPRESS_WP_URL ? process.env.NUXTPRESS_WP_URL : 'http://localhost:3080';

export default {
  debug,
  mode: 'universal',
  env: {
    nuxtDebug: debug,
    nuxtFrontUrl: process.env.NUXTPRESS_FRONT_URL || 'localhost:3000',
    NUXTPRESS_WP_PORT: process.env.NUXTPRESS_WP_PORT || '3080',
    NUXTPRESS_WP_URL: wpUrl,
  },
  /*
   ** Headers of the page
   */
  head: {
    title: process.env.npm_package_name || '',
    meta: [{ charset: 'utf-8' }, { name: 'viewport', content: 'width=device-width, initial-scale=1' }],
    link: [{ rel: 'icon', type: 'image/x-icon', href: '/favicon.ico' }],
  },
  /*
   ** Customize the progress-bar color
   */
  loading: { color: '#683ba3' },
  /*
   ** Global CSS
   */
  css: ['~/assets/scss/app.scss'],
  /*
   ** StyleResources: SCSS to load in all Vue components - don't load any outputted CSS here, just settings + tools
   */
  styleResources: {
    scss: ['~/assets/scss/settings-tools.scss'],
  },
  /*
   ** Plugins to load before mounting the App
   */
  plugins: ['~/plugins/global-components.js', '~/plugins/vue-logger.js', '~/plugins/breakpoints.js'],
  /*
   ** Nuxt.js dev-modules
   */
  buildModules: [
    // Doc: https://github.com/nuxt-community/eslint-module
    '@nuxtjs/eslint-module',
    // Doc:https://github.com/nuxt-community/stylelint-module
    ['@nuxtjs/stylelint-module', { syntax: 'scss' }],
    // Doc: https://github.com/nuxt-community/dotenv-module
    ['@nuxtjs/dotenv', { filename: '.env' }],
    // Doc: https://github.com/nuxt-community/style-resources-module
    '@nuxtjs/style-resources',
  ],
  /*
   ** Nuxt.js modules
   */
  modules: [
    // Doc: https://axios.nuxtjs.org/usage
    '@nuxtjs/axios',
    // '@nuxtjs/pwa',
    '~/modules/static/',
    [
      '~/modules/wp-api/index',
      {
        endpoint: `${wpUrl}/wp-json/`,
      },
    ],
  ],
  /*
   ** Axios module configuration
   ** See https://axios.nuxtjs.org/options
   */
  axios: {},
  /*
   ** Build configuration
   */
  build: {
    ...build,
  },
  generate: {
    routes: () => {
      return axios.get(`${wpUrl}/wp-json/wuxt/v1/generate`).then(({ data }) => data);
    },
  },
};
